alembic_migration/versions/bece9007ab83_add_documents_tags.py

```diff
@@ -0,0 +1,85 @@
+"""Add documents tags
+
+Revision ID: bece9007ab83
+Revises: 06d2a35e39c8
+Create Date: 2019-11-10 16:46:13.660416
+
+"""
+from alembic import op
+from alembic_migration import extensions
+import sqlalchemy as sa
+
+
+# revision identifiers, used by Alembic.
+revision = 'bece9007ab83'
+down_revision = '06d2a35e39c8'
+branch_labels = None
+depends_on = None
+
+view_users_for_routes = extensions.ReplaceableObject(
+      'guidebook.users_for_routes',
+      """
+SELECT guidebook.documents_tags.document_id AS route_id, array_agg(guidebook.documents_tags.user_id) AS user_ids
+FROM guidebook.documents_tags
+GROUP BY guidebook.documents_tags.document_id;
+""")
+
+def upgrade():
+    op.create_table(
+        'documents_tags',
+        sa.Column('user_id', sa.Integer(), nullable=False),
+        sa.Column('document_id', sa.Integer(), nullable=False),
+        sa.Column('document_type', sa.String(1), nullable=False),
```

Comment:
what is the purpose of the document_type ? It seems to me redundant with the information in guidebook.documents

the document_type does not seem to be used in this PR ? can it be removed ?

Reply:
The document type is used by the ES syncer to build its index:
* get the list of documents to update in the index: https://github.com/c2corg/v6_api/blob/b1ec9750e4cf6c47513254bc83381b65d881215f/c2corg_api/scripts/es/sync.py#L35-L40
* type of the changed documents is then used at https://github.com/c2corg/v6_api/blob/b1ec9750e4cf6c47513254bc83381b65d881215f/c2corg_api/scripts/es/sync.py#L275-L279

Except this the document type is as for now indeed not that relevant because all tagged documents are routes ("todo" tag). But the goal of this PR is to be extended for other kind of documents with other kind of tags.

Reply:
OK, thanks for your explanation.
still the data is redundant, couldn't it be handled with an sql **view**, I don't feel comfortable with adding the same data once more and risk inconsistent data ?

Reply:
The concern about data inconsistency makes sense indeed.

We could use views or other systems to get the type directly from the document itself but wouldn't it make things even more complicated? Perhaps we could make sure the type is correctly retrieved when creating the tags.

The current PR make sure the document about to be tagged is actually a route, see https://github.com/c2corg/v6_api/blob/routes-todo-list/c2corg_api/views/document_tag.py#L32-L42

Please also note the document tags system is significantly inspired from the association system (dedicated table for the the actual associations + a table for the logs, as for the tags system). Associations do have document type fields duplicated in the association tables:
https://github.com/c2corg/v6_api/blob/master/c2corg_api/models/association.py#L45-L52

Reply:
ah, that's why. Yes associations also have the document type. Actually I started working on #804 and already forgot about the document types and potential inconsistency.

I checked in the db, it is really easy to retrieve the doc type directly from the table guidebook.documents (without further joins,...) so it's probably not much more complicated to retrieve the document type on the fly. Performance-wise, the foreign key is indexed, so it should be OK ?

Reply:
ok, I reindexed ES locally, and now also the `routes?u=` entrypoint is working (I tested it manually, but I'll look also into the test written by @cbeauchesne )